File: test/floydwarshall.cpp
#include <Helium/algorithms/floydwarshall.h>
#include <Helium/smiles.h>
#include <Helium/fileio/moleculefile.h>

#include "test.h"

using namespace Helium;

int main()
{
  typedef molecule_traits<HeMol>::atom_type atom_type;

  HeMol mol;
  atom_type a1 = mol.addAtom();
  atom_type a2 = mol.addAtom();
  atom_type a3 = mol.addAtom();
  atom_type a4 = mol.addAtom();
  atom_type a5 = mol.addAtom();
  atom_type a6 = mol.addAtom();
  atom_type a7 = mol.addAtom();

  mol.addBond(a1, a2);
  mol.addBond(a2, a3);
  mol.addBond(a3, a4);
  mol.addBond(a4, a5);
  mol.addBond(a5, a6);
  mol.addBond(a6, a7);
  mol.addBond(a7, a1);
  mol.addBond(a2, a6);

  DistanceMatrix D = floyd_warshall(mol);

  COMPARE(0, D(0, 0));
  COMPARE(1, D(0, 1));
  COMPARE(2, D(0, 2));
  COMPARE(3, D(0, 3));
  COMPARE(4, D(0, 4));
  COMPARE(2, D(0, 5));
  COMPARE(3, D(0, 6));

  std::cout << D << std::endl;
}

File: test/python/floydwarshall.py
import helium
import unittest

class TestFloydWarshall(unittest.TestCase):

    def test_floyd_warshall(self):
        mol = helium.Molecule()

        a1 = mol.addAtom()
        a2 = mol.addAtom()
        a3 = mol.addAtom()
        a4 = mol.addAtom()
        a5 = mol.addAtom()
        a6 = mol.addAtom()
        a7 = mol.addAtom()

        mol.addBond(a1, a2)
        mol.addBond(a2, a3)
        mol.addBond(a3, a4)
        mol.addBond(a4, a5)
        mol.addBond(a5, a6)
        mol.addBond(a6, a7)
        mol.addBond(a7, a1)
        mol.addBond(a2, a6)

        D = helium.floyd_warshall(mol)
        self.assertTrue(isinstance(D, helium.DistanceMatrix))
        self.assertEqual(7, D.dim())
        self.assertEqual(0, D(0, 0))
        self.assertEqual(1, D(0, 1))
        self.assertEqual(2, D(0, 2))
        self.assertEqual(3, D(0, 3))
        self.assertEqual(4, D(0, 4))
        self.assertEqual(2, D(0, 5))
        self.assertEqual(3, D(0, 6))

        self.assertTrue(isinstance(D.infinity(), int))


if __name__ == '__main__':
    unittest.main()
